utils/alignments.py

```diff
@@ -0,0 +1,163 @@
+    durs = tf.constant(filled_durations, dtype=tf.int32)[tf.newaxis, :]
+    totlen = tf.math.reduce_sum(durs) * tf.ones(tf.shape(durs)[1], dtype=tf.int32)[tf.newaxis, :]
+    pos = tf.math.cumsum(tf.concat([[0], durs[-1][:-1]], axis=0))[tf.newaxis, :]
+    ctc = tf.transpose(tf.concat([durs, totlen, pos], axis=0))
+    return tf.map_fn(map_to_padded_length, ctc, dtype=tf.float32).numpy()
+
+
+def clean_attention(binary_attention, jump_threshold):
+    phon_idx = 0
+    clean_attn = np.zeros(binary_attention.shape)
+    for i, av in enumerate(binary_attention):
+        next_phon_idx = np.where(av == av.max())[0]
+        if abs(next_phon_idx - phon_idx) > jump_threshold:
+            next_phon_idx = phon_idx
+        phon_idx = next_phon_idx
+        clean_attn[i, min(phon_idx, clean_attn.shape[0] - 1)] = 1
+    return clean_attn
+
+
+def weight_mask(attention_weights):
+    max_m, max_n = attention_weights.shape
+    I = np.tile(np.arange(max_n), (max_m, 1)) / max_n
+    J = np.swapaxes(np.tile(np.arange(max_m), (max_n, 1)), 0, 1) / max_m
+    return np.sqrt(np.square(I - J))
+
+
+def fill_zeros(duration):
+    for i in range(len(duration)):
+        if i < (len(duration) - 1):
+            if duration[i] == 0:
+                next_avail = np.where(duration[i:] > 1)[0]
+                if len(next_avail) > 1:
+                    next_avail = next_avail[0]
+                if next_avail:
+                    duration[i] = 1
+                    duration[i + next_avail] -= 1
+    return duration
+
+
+def compute_cleanings(binary_attn, alignments_weights, binary_score):
+    clean_scores = []
+    clean_attns = []
+    for jumpth in [2, 3, 4, 5]:
+        cl_at = clean_attention(binary_attention=binary_attn, jump_threshold=jumpth)
+        clean_attns.append(cl_at)
+        sclean_score = np.sum(alignments_weights * cl_at)
+        clean_scores.append(sclean_score)
+    best_idx = np.argmin(clean_scores)
+    best_score = clean_scores[best_idx]
+    best_cleaned_attention = clean_attns[best_idx]
+    while ((best_score - binary_score) > 2.) and (jumpth < 11):
+        jumpth += 1
+        best_cleaned_attention = clean_attention(binary_attention=binary_attn, jump_threshold=jumpth)
+        best_score = np.sum(alignments_weights * best_cleaned_attention)
+    phoneme_duration = best_cleaned_attention.sum(axis=0)
+    filled_durations = fill_zeros(phoneme_duration)  # remove zero durations
+    return filled_durations, best_cleaned_attention, best_score, jumpth
+
+
+def binary_attention(attention_weights):
+    attention_peak_per_phoneme = attention_weights.max(axis=1)
+    binary_attn = (attention_weights.T == attention_peak_per_phoneme).astype(int).T
+    binary_score = np.sum(attention_weights * binary_attn)
+    return binary_attn, binary_score
+
+
+def get_durations_from_alignment(batch_alignments, mels, phonemes, weighted=True, PLOT_OUTLIERS=False, PLOT_ALL=False):
+    mel_pad_mask = create_mel_padding_mask(mels)
```

Comment:
This is pretty compliated, I think it would be beneficial to have some comments.